File: 2024/day11/python/solution.py
from pathlib import Path
from functools import lru_cache

input_text = (Path(__file__).parent.parent / "input.txt").read_text().strip()

# Parse input - space-separated numbers
stones = list(map(int, input_text.split()))


@lru_cache(maxsize=None)
def count_stones(value: int, blinks: int) -> int:
    """Count how many stones result from a single stone after N blinks."""
    if blinks == 0:
        return 1

    # Rule 1: 0 becomes 1
    if value == 0:
        return count_stones(1, blinks - 1)

    # Rule 2: Even number of digits -> split
    s = str(value)
    if len(s) % 2 == 0:
        mid = len(s) // 2
        left = int(s[:mid])
        right = int(s[mid:])
        return count_stones(left, blinks - 1) + count_stones(right, blinks - 1)

    # Rule 3: Multiply by 2024
    return count_stones(value * 2024, blinks - 1)


def part1():
    return sum(count_stones(stone, 25) for stone in stones)


def part2():
    return sum(count_stones(stone, 75) for stone in stones)


if __name__ == "__main__":
    print(f"Part 1: {part1()}")
    print(f"Part 2: {part2()}")

File: 2024/day11/ruby/solution.rb
#!/usr/bin/env ruby

# Read input from ../input.txt
input_path = File.join(File.dirname(__FILE__), '..', 'input.txt')
input_text = File.read(input_path).strip

# Parse input - space-separated numbers
stones = input_text.split.map(&:to_i)

# Memoization cache for count_stones
$memo = {}

# Count how many stones result from a single stone after N blinks
def count_stones(value, blinks)
  # Base case: no more blinks
  return 1 if blinks == 0

  # Check memo cache
  key = [value, blinks]
  return $memo[key] if $memo.key?(key)

  result = if value == 0
    # Rule 1: 0 becomes 1
    count_stones(1, blinks - 1)
  else
    s = value.to_s
    if s.length.even?
      # Rule 2: Even number of digits -> split in half
      mid = s.length / 2
      left = s[0...mid].to_i
      right = s[mid..-1].to_i
      count_stones(left, blinks - 1) + count_stones(right, blinks - 1)
    else
      # Rule 3: Multiply by 2024
      count_stones(value * 2024, blinks - 1)
    end
  end

  $memo[key] = result
  result
end

def part1(stones)
  stones.sum { |stone| count_stones(stone, 25) }
end

def part2(stones)
  stones.sum { |stone| count_stones(stone, 75) }
end

puts "Part 1: #{part1(stones)}"
puts "Part 2: #{part2(stones)}"
